text/profiles/0239-profiles-data-model.md

```diff
@@ -0,0 +1,1604 @@
+                - 2
+              sample:
+                - locations_start_index: 0
+                  locations_length: 3
+                  timestamps:
+                    - 1687841528000000
+                  value:
+                    - 100
+                - locations_start_index: 0
+                  locations_length: 2
+                  value:
+                    - 200
+              string_table:
+                - ""
+                - foo
+                - bar
+                - baz
+                - cpu
+                - samples
+```
+
+### Notable Differences Compared to Other Signals
+
+Due to the increased performance requirements associated with profiles signal, here are some notable differences between profiles signal and other signals.
+
+#### Relationships Between Messages
+
+There are two main ways relationships between messages are represented:
+
+* by embedding a message into another message (standard protobuf way)
+* by referencing a message by index (similar to how it's done in pprof)
+
+Profiling signal is different from most other ones in that we use the referencing technique a lot to represent relationships between messages where there is a lot of duplication happening. This allows to reduce the size of the resulting protobuf payload and the number of objects that need to be allocated to parse such payload.
+
+This example illustrates the conceptual difference between the two approaches. Note that this example is simplified for clarity and provided in YAML format for legibility:
+
+```yaml
+# denormalized
+samples:
+- stacktrace:
+  - foo
+  - bar
+  value: 100
+  attribute_set:
+    endpoint: "/v1/users"
+- stacktrace:
+  - foo
+  - bar
+  - baz
+  value: 200
+  attribute_set:
+    endpoint: "/v1/users"
+
+# normalized
+attribute_sets:
+- endpoint: "/v1/users"
+samples:
+- stacktrace:
+  - foo
+  - bar
+  value: 100
+  attribute_set_index: 0
+- stacktrace:
+  - foo
+  - bar
+  - baz
+  value: 200
+  attribute_set_index: 0
+
+```
+
+Explanation: because multiple samples have the same attributes, we can store them in a separate table and reference them by index. This reduces the size of the resulting protobuf payload and the number of objects that need to be allocated to parse such payload.
+
+Benchmarking shows that this approach is significantly more efficient in terms of CPU utilization, memory consumption and size of the resulting protobuf payload. See [Prior art and alternatives](#prior-art-and-alternatives) for more details.
+
+#### Relationship Between Samples and Locations
+
+Relationship between Samples and Locations is using the referencing technique described above. However, there's an additional optimization technique used to reduce the size of the resulting protobuf payload and the number of objects that need to be allocated to parse such payload. The technique is based on the fact that many samples share the same locations.
+
+Considering the following example presented in a folded format:
+
+```
```

Comment:
Adding two more stacks to this example, so in the folded format it looks like this:
```
foo;bar;baz 100
abc;def 200
foo;bar 300
abc;ghi 400
foo;bar;qux 500
```
Like in most stack traces, the base frames are similar, but there is a variation in the leaf frames. To reflect this, the two added traces use different leaf frames, `ghi` and `qux`.

Should the resulting `sample` look like the following?
```yaml
sample:
  - locations_start_index: 0
    locations_length: 3
    value:
      - 100
  - locations_start_index: 3
    locations_length: 2
    value:
      - 200
  - locations_start_index: 0
    locations_length: 2
    value:
      - 300
  - locations_start_index: 5
    locations_length: 2
    value:
      - 400
  - locations_start_index: 7
    locations_length: 3
    value:
      - 500
location_indices:
  - 0 # foo
  - 1 # bar
  - 2 # baz
  - 3 # abc
  - 4 # def
  - 3 # abc 
  - 5 # ghi
  - 0 # foo
  - 1 # bar
  - 6 # qux
location:
  - line:
      - function_index: 0 # foo
  - line:
      - function_index: 1 # bar
  - line:
      - function_index: 2 # baz
  - line:
      - function_index: 3 # abc
  - line:
      - function_index: 4 # def
  - line:
      - function_index: 5 # ghi
   - line:
      - function_index: 6 # qux
function:
  - name: 1 # foo
  - name: 2 # bar
  - name: 3 # baz
  - name: 4 # abc
  - name: 5 # def
  - name: 6 # ghi
  - name: 7 # qux
```

In particular for deep stack traces with a high number of similar frames and where only leaf frames are different, the use of `locations_start_index`, `locations_length` with `location_indices` will get more complex than the (deprecated) `location_index` which just holds a list of IDs into the location table.

Reply:
The `location_indices` approach is primarily about reusing identical stacks between samples (e.g. when only labels are different) in a way that does not incur too many additional allocations. It is not an attempt to optimize encoding shared prefixes of a call tree. Neither previous nor current approach do that efficiently.

Reply:
The original [pprof message Sample](https://github.com/google/pprof/blob/fb44976bdcd59b48559fb1763d69313508ad3a75/proto/profile.proto#L104-L128) does also not use the `_start_index` / `_length` approach. From my understanding all messages of type `Sample` within the same `Profile` groups stack traces from the same origin/with the same attributes.
For a different set of attributes, I think, a dedicated `Profile` should be preferred with its own attributes.

An alternative, to allow sharing `Mapping`, `Location` and `Function` information between stack traces with different attributes would be to move these three tables one layer up into `ProfileContainer`, so that they can be referenced from each `Profile`.

While the variety of leaf frames is usually high and attributes are often more static, can we remove the `deprecated` label from `location_index` in `message Sample` and let the user either set `location_index` or `location_start_index` with `locations_length`?

Reply:
Discussed with profiling SiG.  We believe this discussion can continue as the model is implemented in the proto directory, and does not need to be addressed in the proposal, as it doesn't change overall direction.